### ISSUE TITLE
BUILD_CAFFE2=0 causes build failure in some configurations

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#45600]
Here is my build command, which basically turns all extra stuff off for the fastest build:
```
DEBUG=1 BUILD_TEST=0 BUILD_CAFFE2_OPS=0 USE_OPENMP=0 USE_MKL=0 USE_MKLDNN=0 USE_NNPACK=0 USE_XNNPACK=0 USE_QNNPACK=0 USE_PYTORCH_QNNPACK=0 USE_FBGEMM=0 USE_DISTRIBUTED=0 USE_CUDA=0 python setup.py -q develop
```
I didn't look into it too closely, but it seems `BUILD_CAFFE2=0` is dropping compiling the pthreadpool dependency which we need. This seems to be covered up by the fact that NNPack and family are adding back the threadpool dependency, but if you turn those off as well (and thus `DISABLE_NN_PACK_AND_FAMILY=1`), then we're left with a missing dependency. 

I guess we need to unconditionally compile `pthreadpool`?

If I add `BUILD_CAFFE2=0` to this big mix, I get the following error:
```
FAILED: caffe2/CMakeFiles/torch_cpu.dir/utils/threadpool/pthreadpool-cpp.cc.o 
/usr/local/bin/clang++ -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DUSE_EXTERNAL_MZCRC -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I../torch/csrc/api -I../torch/csrc/api/include -I../caffe2/aten/src/TH -Icaffe2/aten/src/TH -Icaffe2/aten/src -Icaffe2/../aten/src -Icaffe2/../aten/src/ATen -I../torch/csrc -I../third_party/miniz-2.0.8 -I../aten/src/TH -I../aten/../third_party/catch/single_include -I../aten/src/ATen/.. -Icaffe2/aten/src/ATen -I../c10/.. -I../third_party/cpuinfo/include -I../third_party/FP16/include -I../third_party/fmt/include -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /home/suo/.conda/envs/pytorch-dev/include -isystem ../cmake/../third_party/eigen -isystem /home/suo/.conda/envs/pytorch-dev/include/python3.8 -isystem /home/suo/.conda/envs/pytorch-dev/lib/python3.8/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem include -fuse-ld=lld -Wno-deprecated -fvisibility-inlines-hidden -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -g -fno-omit-frame-pointer -O0 -fPIC -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -DCAFFE2_BUILD_MAIN_LIB -pthread -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/utils/threadpool/pthreadpool-cpp.cc.o -MF caffe2/CMakeFiles/torch_cpu.dir/utils/threadpool/pthreadpool-cpp.cc.o.d -o caffe2/CMakeFiles/torch_cpu.dir/utils/threadpool/pthreadpool-cpp.cc.o -c ../caffe2/utils/threadpool/pthreadpool-cpp.cc
/usr/bin/clang++: /lib64/libtinfo.so.6: no version information available (required by /usr/bin/clang++)
../caffe2/utils/threadpool/pthreadpool-cpp.cc:6:1: error: use of undeclared identifier 'PThreadPool'
PThreadPool::PThreadPool(const size_t thread_count)
^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:9:8: error: use of undeclared identifier 'PThreadPool'
size_t PThreadPool::get_thread_count() const {
       ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:16:6: error: use of undeclared identifier 'PThreadPool'
void PThreadPool::set_thread_count(const size_t thread_count) {
     ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:28:6: error: use of undeclared identifier 'PThreadPool'
void PThreadPool::run(
     ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:29:16: error: no template named 'function' in namespace 'std'
    const std::function<void(size_t)>& fn,
          ~~~~~^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:58:1: error: unknown type name 'PThreadPool'
PThreadPool* pthreadpool() {
^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:59:15: error: no template named 'unique_ptr' in namespace 'std'
  static std::unique_ptr<PThreadPool> threadpool =
         ~~~~~^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:59:26: error: use of undeclared identifier 'PThreadPool'; did you mean 'pthreadpool'?
  static std::unique_ptr<PThreadPool> threadpool =
                         ^~~~~~~~~~~
                         pthreadpool
../caffe2/utils/threadpool/pthreadpool-cpp.cc:58:14: note: 'pthreadpool' declared here
PThreadPool* pthreadpool() {
             ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:60:12: error: no member named 'make_unique' in namespace 'std'
      std::make_unique<PThreadPool>(getDefaultNumThreads());
      ~~~~~^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:60:24: error: use of undeclared identifier 'PThreadPool'
      std::make_unique<PThreadPool>(getDefaultNumThreads());
                       ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:64:1: error: unknown type name 'pthreadpool_t'
pthreadpool_t pthreadpool_() {
^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:65:3: error: use of undeclared identifier 'PThreadPool'; did you mean 'pthreadpool'?
  PThreadPool* const threadpool = pthreadpool();
  ^~~~~~~~~~~
  pthreadpool
../caffe2/utils/threadpool/pthreadpool-cpp.cc:58:14: note: 'pthreadpool' declared here
PThreadPool* pthreadpool() {
             ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:65:16: error: expected expression
  PThreadPool* const threadpool = pthreadpool();
               ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:67:7: error: use of undeclared identifier 'threadpool'; did you mean 'pthreadpool'?
      threadpool, "Failed to acquire an instance of PThreadPool!");
      ^~~~~~~~~~
      pthreadpool
../c10/util/Exception.h:285:31: note: expanded from macro 'TORCH_INTERNAL_ASSERT'
  if (C10_UNLIKELY_OR_CONST(!(cond))) {       \
                              ^
../c10/util/Exception.h:238:47: note: expanded from macro 'C10_UNLIKELY_OR_CONST'
#define C10_UNLIKELY_OR_CONST(e) C10_UNLIKELY(e)
                                              ^
../c10/macros/Macros.h:171:65: note: expanded from macro 'C10_UNLIKELY'
#define C10_UNLIKELY(expr)  (__builtin_expect(static_cast<bool>(expr), 0))
                                                                ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:58:14: note: 'pthreadpool' declared here
PThreadPool* pthreadpool() {
             ^
../caffe2/utils/threadpool/pthreadpool-cpp.cc:68:10: error: use of undeclared identifier 'threadpool'
  return threadpool->threadpool_.get();
         ^
15 errors generated.
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "setup.py", line 720, in <module>
    build_deps()
  File "setup.py", line 311, in build_deps
    build_caffe2(version=version,
  File "/home/suo/code/pytorch/tools/build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "/home/suo/code/pytorch/tools/setup_helpers/cmake.py", line 345, in build
    self.run(build_args, my_env)
  File "/home/suo/code/pytorch/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/home/suo/.conda/envs/pytorch-dev/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Debug', '--', '-j', '12']' returned non-zero exit status 1.
```

cc @malfet @seemethere @walterddr

**Comment by suo** [comment]
cc @malfet  @walterddr

**Comment by suo** [comment]
should be fixed by https://github.com/pytorch/pytorch/pull/45610

**Comment by soCzech** [comment]
The issue does not seem to be fixed by the commit as it occurs in pytorch v1.7.1

The cmake setup command:
```
cmake -DTORCH_CUDA_ARCH_LIST="6.1" -DBUILD_SHARED_LIBS=OFF -DBUILD_CUSTOM_PROTOBUF=OFF -DUSE_FBGEMM=OFF -DUSE_QNNPACK=OFF -DUSE_PYTORCH_QNNPACK=OFF -DUSE_NNPACK=OFF -DUSE_XNNPACK=OFF -DUSE_DISTRIBUTED=OFF -DUSE_MKLDNN=OFF -DBUILD_PYTHON=OFF -DBUILD_CAFFE2=OFF -DUSE_CUDA=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr/local/libtorch ${DIR}/pytorch
```

**Comment by malfet** [comment]
Updated, but that's a blast from the past. Closing, as majority of PyTorch public builds are done with `BUILD_CAFFE2=0` and they are working just fine.
